src/core/exec-invoke.c

```diff
@@ -4873,6 +5001,32 @@ int exec_invoke(
                 }
         }
 
+        /* Unshare a new PID namespace before setting up mounts to ensure /proc is mounted with only processes in PID namespace visible.
+         * Note PrivatePIDs=yes implies MountAPIVFS=yes so we'll always ensure procfs is remounted. */
+        if (needs_sandboxing && exec_needs_pid_namespace(context) && params->pidref_transport_fd >= 0) {
+                r = can_mount_proc(context, params);
+                if (r > 0) {
+                        r = setup_private_pids(context, params);
+                        if (ERRNO_IS_NEG_PRIVILEGE(r))
+                                log_exec_warning_errno(context, params, r,
+                                                        "PrivatePIDs=yes is configured, but pid namespace setup failed due to lack of privileges, ignoring: %m");
+                        else if (r < 0) {
+                                *exit_status = EXIT_NAMESPACE;
+                                return log_exec_error_errno(context, params, r, "Failed to set up pid namespace: %m");
+                        }
+
+                        if (r > 0)
+                                /* Return value > 0 means we are in the parent process. Thus, return immediately in order to exit ASAP.
+                                * The rest of the setup will continue in the child process. */
+                                return 0;
+                } else if (r == 0)
+                        log_exec_warning(context, params, "PrivatePIDs=yes is configured, but /proc cannot be re-mounted due to lack of privileges, ignoring.");
+                else
+                        log_exec_warning_errno(context, params, r, "PrivatePIDs=yes is configured, but failed to detect if /proc can be remounted, ignoring: %m");
+        }
```

Comment:
so I think that failure to get PrivatePIDs=yes should fail the service. Most security related options we handle gracefully, under the assumption that not applying them just gives you more rights but won't change execution in a meaningful way. But I think for PrivatePIDs= that#s not the case. It very much changes execution, because PID 1 has quite different semantics from other PIDs (they are target for reparenting and such), moreover things like writing pid files (for those legacy daemons that do that) becomes really wweird, because what they'd write there might not match what the rest of the system sees.

if you want this to be graceful, a middle ground might be to add PrivatePIDs=try or so, so that people explicitly opt into graceful handling of this knob. But if you don#t need this I wouldn't bother and just make inability to apply privatepids a failure here.

Compare how BindPath= failing will casue a service failure for us, but ReadOnlyPath= is handled gracefully. That also follows the logic that the former rearrranges things too heavily to ignore if it doesnt work. The latter only takes away privs.

Reply:
Yea at least for me, we don't need/want graceful fallback. So I will remove it.